python-package/xgboost/spark/core.py

```diff
@@ -780,14 +781,21 @@ def _train_booster(pandas_df_iter):
             else:
                 use_qdm = use_hist
 
+            _rabit_args = {}
+            if context.partitionId() == 0:
+                _rabit_args = _get_rabit_args(context, num_workers)
+
+            worker_message = {
+                "rabit_msg": _rabit_args,
+                "use_qdm": use_qdm,
+            }
+
+            messages = context.allGather(message=json.dumps(worker_message))
+
+            use_qdm = all(json.loads(x)["use_qdm"] for x in messages)
```

Comment:
All workers should be consistent. Let's use a check instead of a workaround.

Reply:
We can't assume the user's env, especially for standalone mode. users need to install the xgboost dep in every worker manually. And this check is really cheap since the message is attached to the original allGather API getting rabit message.  but it can really fix the "inconsistent" issue

Reply:
And this PR also removes the first barrier() which is not needed, since allGather achieves the same goal.

Reply:
We would like the users to have a consistent environment.

> users need to install the xgboost dep in every worker manually

I'm sure a user can automate the process.

My suggestion is not to remove the allgather, but instead to **assert** that all workers have the same environment.

Reply:
what do you mean "not to remove the allgather"?

@WeichenXu123 what's your opinion about *but instead to assert that all workers have the same environment. If it doesn't, we will throw an exception and terminate the job.*

Reply:
@WeichenXu123, how about throwing an exception when detecting in-consistent env?

Reply:
The code is simple and won't introduce any side-effects including extra overhead.

Reply:
@wbo4958 Sounds good!